Preview of interface work

superbudvar opened the pull request:
Hi Guys!

This is far from complete, but I thought I should submit a work in progress for you to review. 
It's all up-to-date, and while buggy, doesn't seem to be "crash" buggy (the keyboard glitch is really annoying me)

I've created a common.h because of linker problems; no idea if that's the right way to go, but I wanted it working sooner rather than later so you can see it and I know whether to give up pursuing this route or not

There are more comments in the commit log

Any questions/comments appreciated.

superbudvar commented:
oh, hope you've got citro3d installed; another dependancy now

llakssz commented:
Nice I'm going to have a look at this in a little while, excited!

DrakiaXYZ commented:
I'll check this out after work! :)

DrakiaXYZ commented:
Commit 330c5d0 is going to break building for other people, as it hard-codes the path to buildtools outside the project directory.

DrakiaXYZ commented on source/main.cpp:
```diff
@@ -355,30 +361,36 @@ std::string getInput(HB_Keyboard* sHBKB, bool &bCancelled)
+                sceneDraw();
+                last_input = input;
+            }
+            //Wait for VBlank
```
We may need to do something about these. HBKB doesn't draw properly without them (Some buttons get stuck with a weird flickery blue color to them). Not sure how to handle it though, as with these two lines in, the top screen goes flickery while typing.

superbudvar replied:
Yes; it's one of my pet hates right now. I need to look into the routines around it a bit more. I assumed that maybe I hadn't switched to the correct video mode or something similar.

DrakiaXYZ commented on source/menu.cpp:
```diff
@@ -21,17 +22,34 @@ void init_menu(gfxScreen_t screen)
+    setTextColor(0xFF00FF00);
+    renderText(0, 8, 0.7f, 0.7f, false, title);
+    for (int i = 0; i < count && i < (currentMenu.menuConsole.consoleHeight - 2); i++) {
+        ui_menu_draw_string(options[i], 1, 32+(i*12), 0xFF000000);
```
Not really a fan of the magic numbers here (32+(i*12)). Are these the top offset for the menu, and the height of lines? Is there some way to dynamically get them?

superbudvar replied:
lol; yes they are guesses as to offsets and height. The renderText function (as re-appropriated from 3ds-examples) does lots of clever glyph width and height calculations. I was thinking this could be re-aligned to your currentMenu x and y positions and multiplied internally, but then issues arise from the x value as the font is variable width.

I chose not to get too hung up on it, as going forward, as I would image this to be a stop-gap really. I agree swapping out for defines will help readability

DrakiaXYZ commented on source/main.cpp:
```diff
@@ -413,6 +425,12 @@ std::istream& GetLine(std::istream& is, std::string& t)
     }
 }
 
+u64 hex_to_u64( std::string value) {
```
Is this used anywhere?

superbudvar replied:
oops not yet, it's used for converting the titleId (string) to a u64 for passing, mainly to getProductCode to see if the title is installed or not.

DrakiaXYZ commented on source/menu.cpp:
```diff
@@ -62,57 +81,45 @@ void titles_multkey_draw(const char *title, const char* footer, int back, std::v
+            menu_draw_string(title, 0, current_pos_y++, CONSOLE_RED);
+            menu_pos_y = current_pos_y;
+            for (int i = 0; menu_offset + i < count && i < results_per_page; i++) {
+                int y_pos = 18+(i*12);
```
Some more magic numbers in this function, along with menu_end_y above which used to be based on console height. Hopefully we can make these dynamically determined :)

DrakiaXYZ commented:
Left a couple of comments inline above :) Some other things I noticed:
- indentation is a bit messed up in a few places (Mixed spaces and tabs)
- using inline colors, I see you defined some in the header, assuming you just haven't gotten around to cleaning up the inline stuff?

Looks great so far though :+1: Let me know if you want help with anything on it :)

superbudvar commented:
Thanks for the feedback! :+1: 

The buildtools part of the commit is my bad; Shouldn't be there as that just what I had to do to get things building :smirk: 

Yes; it's very much work in progress. I've switched my ide to use spaces rather than tabs, but probably a bit too late.

I've definitely not gotten to a tidy up phase, as I had a crisis as to whether it was the best way to go with things and wanted to check you guys were happy with it's direction. I have got defines for colours locally now, and done a refactor; mainly because I couldn't work out how to call stuff in main.cpp from externals, but I also think it'll help reduce merge conflicts also.

I'd love to have some help with this; I'm not proud :laughing: I'll push my recent work, but I think the next thing needed is to determine what needs to be done from here to make it a release candidate; I'm quite happy with the output of menus to be on the bottom where they haven't been "converted" but then I wonder if we should out and out replace the console too

superbudvar commented:
Hello.

Some updates, for your consideration. Comments again please. I've addressed most of the comments so far

keyboard is still glitchly however and there are still a few magic offsets and some routines like download and installs still fall to the lower screen but I think it's in a lot better shape.

DrakiaXYZ commented:
I'll take a look at this this weekend :+1:

DrakiaXYZ commented on source/DownloadQueue.h:
```diff
@@ -0,0 +1,31 @@
+
+#include "utils2.h"
+#include "utils.h"
+#include "ticket.h"
```
I'd recommend keeping header includes in the .c/.cpp file unless they're actually used in the header itself, just makes things cleaner :)

superbudvar replied:
Ah ok. For some reason, I thought best practice was to "hide" these in the .h

I didn't like it myself tbh

DrakiaXYZ commented on source/main.cpp:
```diff
@@ -783,12 +475,13 @@ void action_about()
 
-    printf(CONSOLE_RED "CIAngel by cearp and Drakia\n" CONSOLE_RESET);
+    printf(CONSOLE_RED "CIAngel by cearp and Drakia\n");
+    printf( CONSOLE_RESET);
```
Not a fan of this particular change, the old code showed the intention better (Change to red, output text, change back), whereas this one someone may see the CONSOLE_RESET on its own and delete it by accident

superbudvar replied:
This is an interesting one, actually.. the reason for this change was to stop the app from crashing! :open_mouth: 

Not sure why, but this change actually made a difference.

DrakiaXYZ commented:
- Doesn't look like there's any output when you hit "Download wings.json" until it's reloading it
- Getting some flicker on the bottom screen status bar, nothing too bad but I know someone will bring it up eventually
- Getting last search screen flicker when going into search more than once (Probably due to double buffering, maybe there's a way to clear the screen buffer before we show results or before we draw the scene, haven't looked too much into it yet)
- "Title/Ticket installed" shows up when svcHax isn't available, and just always says no
- Closing the 3DS while downloading results in a freeze (It's so easy to re-introduce this bug)

Looking pretty good :+1:

airkingbr commented:
nice work! Very beautiful new interface! Thanks !

superbudvar commented:
Thanks! Hoping to get it nailed and released soon!
